// hhdoctor-wmp-sdk/src/hhBehaviors.js
module.exports = Behavior({
  behaviors: [],
  properties: {
    profileName: {
      type: String,
      value: 'prod'
    },

    sdkProductId: {
      type: String,
      observer(newVal, oldVal, changedPath) {
        this.propertyChanged(newVal, oldVal, changedPath);
      }
    },
    appKey: {
      type: String,
      value: ''
    },
    userUuid: {
      type: String,
      observer(newVal, oldVal, changedPath) {
        this.propertyChanged(newVal, oldVal, changedPath);
      }
    },
    userToken: {
      type: String,
      observer(newVal, oldVal, changedPath) {
        this.propertyChanged(newVal, oldVal, changedPath);
      }
    },
    openId: {
      type: String,
      observer(newVal, oldVal, changedPath) {
        this.propertyChanged(newVal, oldVal, changedPath);
      }
    }
  },
  data: {},
  attached() {},
  methods: {
    propertyChanged(newVal, oldVal, changedPath) {},

    getHost() {
      //wsServer: 'wss://wmp.hh-medic.com/wmp/websocket',
      //fileServer: 'https://dev.hh-medic.com/miniprogramweb_master/wmp/im/upload/'
      var host = {};
      switch (this.data.profileName) {
        case 'prod':
          host.wmpHost = 'https://wmp.hh-medic.com/wmp/';
          host.ehrHost = 'https://e.hh-medic.com/ehrweb/';
          host.wsServer ='wss://wmp.hh-medic.com/wmp/websocket';
          break;
        case 'test':
          host.wmpHost = 'https://test.hh-medic.com/wmp/';
          host.ehrHost = 'https://test.hh-medic.com/ehrweb/';
          host.wsServer = 'wss://test.hh-medic.com/wmp/websocket';
          break;
        default:
          break;
      }
      return host;
    }
  }
})
